torch/nn/attention/flex_attention.py

```diff
@@ -1154,6 +1154,61 @@ def _validate_nestedness(query: Tensor, key: Tensor, value: Tensor):
         )
 
 
+def _enforce_mem_layouts(
+    query: Tensor, key: Tensor, value: Tensor
+) -> tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
+    """
+    Enforce memory layouts for query, key, and value tensors.
+
+    For non-FP8 dtypes, no action is taken.
+
+    For FP8 dtypes, we enforce the following memory layouts:
+    - Query tensor must be in row-major memory layout, as it will be the left-operand in the FP8 GEMM `q @ k.T`.
+    - Key tensor must be in row-major memory layout, as it will be transposed when used as the right-operand
+      in the FP8 GEMM `q @ k.T`, meaning it will correctly be in column-major memory layout for the GEMM.
+    - Value tensor must be in column-major memory layout, as it will be the right-operand in the FP8 GEMM `softmax_scores @ v`.
+
+    Returns the query, key, and value tensors with the enforced memory layouts.
+    """
+
+    def is_row_major(tensor: Tensor) -> bool:
+        return tensor.stride()[-1] == 1
+
+    def is_col_major(tensor: Tensor) -> bool:
+        return tensor.stride()[-2] == 1
+
+    # These memory layout constraint are only for FP8 GEMMs on architectures prior to SM100.
+    # SM100 has support for TN, NT, TT, NN layouts for FP8 GEMM
+    # (i.e., left and right operands can be in row or column major layouts)
+    # so this check is only needed for older architectures.
+    # See: https://github.com/NVIDIA/cutlass/blob/main/media/docs/cpp/blackwell_functionality.md
+    fp8_dtypes = (
+        torch.float8_e4m3fn,
+        torch.float8_e5m2,
+    )
+    gemm_precision = query.dtype
+    is_sm100_or_greater = (
+        torch.cuda.is_available()
+        and torch.cuda.get_device_capability("cuda") >= (10, 0)
```

Comment:
```suggestion
        torch.version.cuda
        and torch.cuda.get_device_capability("cuda") >= (10, 0)
```

I ran some tests on the AMD machine that I'm on:
```python
>>> torch.cuda.is_available()
True
>>> torch.version.cuda
>>> torch.version.cuda is None
True
>>> torch.version.hip
'6.3.42131-fa1d09cbd'
>>> torch.cuda.get_device_capability("cuda")
(9, 4)
```

So I think you need to check `torch.version.hip` to be able to reliably tell if you're on an AMD machine

Reply:
Thanks for testing this. Seems like could just check if `torch.version.cuda is None`?